src/struct_mv/struct_communication.c

```diff
@@ -762,6 +762,46 @@ hypre_CommTypeSetEntry( hypre_Box           *box,
    return hypre_error_flag;
 }
 
+HYPRE_Complex *
+hypre_StructCommunicationGetBuffer(HYPRE_MemoryLocation memory_location,
+                                   HYPRE_Int            size)
+{
+   HYPRE_Complex *ptr;
+
+#if defined(HYPRE_USING_GPU)
+   if (hypre_GetActualMemLocation(memory_location) != hypre_MEMORY_HOST)
+   {
+      if (size > hypre_HandleStructCommSendBufferSize(hypre_handle()))
+      {
+         HYPRE_Int new_size = 5 * size;
+         hypre_HandleStructCommSendBufferSize(hypre_handle()) = new_size;
+         hypre_TFree(hypre_HandleStructCommSendBuffer(hypre_handle()), memory_location);
+         hypre_HandleStructCommSendBuffer(hypre_handle()) = hypre_CTAlloc(HYPRE_Complex, new_size, memory_location);
+      }
+
+      ptr = hypre_HandleStructCommSendBuffer(hypre_handle());
+   }
+   else
+#endif
+   {
+      ptr = hypre_CTAlloc(HYPRE_Complex, size, memory_location);
+   }
+
+   return ptr;
+}
+
+HYPRE_Int
+hypre_StructCommunicationReleaseBuffer(HYPRE_Complex       *buffer,
+                                       HYPRE_MemoryLocation memory_location)
+{
+   if (hypre_GetActualMemLocation(memory_location) == hypre_MEMORY_HOST)
```

Comment:
Why are you only freeing this if it's host memory? Also, both here and above, why do you need to use hypre_GetActualMemLocation()?

Reply:
These changes (GetBuffer and Release) are just for making the code cleaner, a little bit maybe. The logic is if it's CPU memory we allocate and deallocate every time, while for GPU memory, we allocate once and save it (reallocate 5x more if needed). So ReleaseBuffer is no-op for GPUs.